theano/gof/toolbox.py

```diff
@@ -273,6 +273,8 @@ def on_attach(self, fgraph):
             if hasattr(fgraph, attr):
                 raise AlreadyThere("ReplaceValidate feature is already present"
                                    " or in conflict with another plugin.")
+        self._nodes_removed = set()
```

Comment:
In the method on_detach, do "self._nodes_removed" to help make sure we loose refcount to some nodes. There is some cases where we keep to much of them, so this is just to help make sure we don't make those cases more troublesome.

After that I think the PR is ready to merge. Thanks